org.eclipse.scanning.test/src/org/eclipse/scanning/test/scan/nexus/PosDetectorScanTest.java

```diff
@@ -0,0 +1,225 @@
+import org.eclipse.january.dataset.IDataset;
+import org.eclipse.january.dataset.ILazyDataset;
+import org.eclipse.january.dataset.PositionIterator;
+import org.eclipse.scanning.api.device.AbstractRunnableDevice;
+import org.eclipse.scanning.api.device.IRunnableDevice;
+import org.eclipse.scanning.api.device.IWritableDetector;
+import org.eclipse.scanning.api.device.models.IDetectorModel;
+import org.eclipse.scanning.api.event.scan.DeviceState;
+import org.eclipse.scanning.api.points.IPosition;
+import org.eclipse.scanning.api.scan.PositionEvent;
+import org.eclipse.scanning.api.scan.ScanningException;
+import org.eclipse.scanning.api.scan.event.IPositionListenable;
+import org.eclipse.scanning.api.scan.event.IPositionListener;
+import org.eclipse.scanning.api.scan.models.ScanModel;
+import org.eclipse.scanning.example.detector.PosDetectorModel;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class PosDetectorScanTest extends NexusTest {
+	
+	private IWritableDetector<PosDetectorModel> detector;
+	
+	@Before
+	public void before() throws Exception {
+		PosDetectorModel model = new PosDetectorModel(3);
+		model.setTimeout(100); // TODO remove
+		detector = (IWritableDetector<PosDetectorModel>) dservice.createRunnableDevice(model);
+		assertNotNull(detector);
+	}
+	
+	@After
+	public void after() throws Exception {
+		File parentDir = output.getParentFile();
+		String fileName = output.getName().substring(0, output.getName().indexOf('.'));
+		File outputDir = new File(parentDir, fileName);
+		for (File file : outputDir.listFiles()) {
+//			file.delete();
+		}
+//		outputDir.delete();
+	}
+	
+	@Test
+	public void testPosScan() throws Exception {
+		final int[] scanShape = new int[] { 8, 5 };
+		IRunnableDevice<ScanModel> scanner = createGridScan(detector, output, false, scanShape);
+		// add a the UniqueKeyChecker as a position listener to check the unique key is written at each point
+		((IPositionListenable) scanner).addPositionListener(
+				new UniqueKeyChecker(scanner.getModel().getFilePath(), scanShape));
+		
+		assertScanNotFinished(getNexusRoot(scanner).getEntry());
+		
+		scanner.run(null);
+		
+		checkNexusFile(scanner, scanShape);
+	}
+	
+	private static class UniqueKeyChecker implements IPositionListener {
+		
+		private final int[] scanShape;
+		private final String filePath;
+		private ILazyDataset uniqueKeysDataset;
+		
+		UniqueKeyChecker(String filePath, final int[] scanShape) {
+			this.filePath = filePath;
+			this.scanShape = scanShape;
+		}
+		
+		@Override
+		public void positionMovePerformed(PositionEvent event) throws ScanningException {
+			checkUniqueKeyWritten(event.getPosition());
+		}
+
+		@Override
+		public void positionPerformed(PositionEvent event) throws ScanningException {
+			checkUniqueKeyWritten(event.getPosition());
+		}
+		
+		private IDataset getUniqueKeysDataset() throws ScanningException {
+			try {
+				ILazyDataset uniqueKeysDataset = null;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename "uniqueKeysDataset" which hides the field declared at line 97. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AHiddenFieldCheck)